torch/csrc/autograd/engine.cpp

```diff
@@ -630,6 +631,9 @@ void validate_outputs(
       ss << metadata.device() << " but got " << grad_device;
       AT_ERROR(format_error(ss.str()));
     }
+    // The gradient type should match the forward Tensor one
+    // We should not build graph for Tensors that are not differentiable
+    TORCH_INTERNAL_ASSERT(isDifferentiableType(grad.scalar_type()));
```

Comment:
isn't this already (attempted) to be checked above: `    TORCH_CHECK(isFloatingType(grad.scalar_type()) || (input_is_complex == grad_is_complex));
`

Reply:
The semantic is a bit weird here because, most notably for quantized training, we allow backrpopagating gradient of a different type from the forward Tensor.
The check you mention above is here to make sure that all the casting operations are valid.
This one is to make sure that the final Tensor we get after all the casting is a valid dtype. (it should always be as it should match the forward dtype which has to be differentiable for us to get here).

Reply:
is the quantized training thing documented anywhere?